Introduce rocAL Tensors

fiona-gladwin opened the pull request:
- Introduce tensor.h, tensor.cpp files and corresponding enums.
- Introduce rocALTensor, rocALTensorList and rocALTensorInfo class.

fiona-gladwin committed:
Add tensor files

fiona-gladwin committed:
Minor fix

fiona-gladwin committed:
Revert enum change

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Change dims vector from unsigned to size_t type

fiona-gladwin committed:
Remove or add whiteline spaces

LakshmiKumar23 commented on rocAL/rocAL/include/pipeline/commons.h:
```diff
@@ -81,6 +96,21 @@ enum class DecodeMode
     CPU = 1
 };
 
```
@fiona-gladwin Can we keep this uniform and either "= 1" everywhere or remove it here?

LakshmiKumar23 commented on rocAL/rocAL/include/pipeline/tensor.h:
```diff
@@ -0,0 +1,263 @@
+ */
+vx_enum vx_mem_type(RocalMemType mem);
+
+vx_size tensor_data_size(RocalTensorDataType data_type);
```
@fiona-gladwin write brief for function

rrawther commented on rocAL/rocAL/include/pipeline/tensor.h:
```diff
@@ -0,0 +1,263 @@
+ * @param mem input Rocal type
+ * @return the OpenVX type associated with input argument
+ */
+vx_enum vx_mem_type(RocalMemType mem);
```
why is this function declared here? Please move common function definitions to .h file

fiona-gladwin replied:
The vx_mem_type function is used only in tensor.cpp file. The image.cpp file will be removed in future once we replace with tensor pipeline hence we have declared the API in tensor.h and defined in tensor.cpp file

rrawther commented on rocAL/rocAL/include/pipeline/tensor.h:
```diff
@@ -0,0 +1,263 @@
+vx_size tensor_data_size(RocalTensorDataType data_type);
+
+/*! \brief Holds the information about a rocALTensor */
+class rocALTensorInfo
```
We have image (which is named Image class). If we need to add rocal before tensor, probably we need to add before image as well. Also I prefer rocalTensor. Please don't use rocAL for naming variables inside the implementation

fiona-gladwin replied:
We will change the name of the class to rocalTensor

rrawther commented on rocAL/rocAL/source/pipeline/tensor.cpp:
```diff
@@ -0,0 +1,357 @@
+#include "commons.h"
+#include "tensor.h"
+
+vx_enum vx_mem_type(RocalMemType mem)
```
please format the entire file according to google coding style

rrawther commented on rocAL/rocAL/source/pipeline/tensor.cpp:
```diff
@@ -0,0 +1,357 @@
+        break;
+        case RocalTensorDataType::FP16:
+        {
+            return sizeof(vx_int16);
```
this will get changed to vx_float16. There is another PR to fix that in amd OpenVX

rrawther commented on rocAL/rocAL/source/pipeline/tensor.cpp:
```diff
@@ -0,0 +1,357 @@
+
+bool operator==(const rocALTensorInfo &rhs, const rocALTensorInfo &lhs)
+{
+    return (rhs.dims() == lhs.dims() &&
```
the value of dimensions are not checked here

fiona-gladwin replied:
When we compare two vectors(the dims) internally the values will also be compared.

rrawther commented on rocAL/rocAL/include/pipeline/tensor.h:
```diff
@@ -0,0 +1,263 @@
+        _data_size = (_data_size / _data_type_size);
+        _data_size *= data_type_size();
+    }
+    void set_max_dims()
```
what is max_dims here? If this is width and height, please rename them as width and height

fiona-gladwin replied:
In the case of image tensors we will store the max width and max height, but when it comes to audio tensors we will be storing the maximum frames and channels, hence we have kept a generic name as max_dims

rrawther commented on rocAL/rocAL/include/pipeline/tensor.h:
```diff
@@ -0,0 +1,263 @@
+                _max_dims[0] = _dims.at(2);
+                _max_dims[1] = _dims.at(1);
+            }
+            else if(_layout == RocalTensorlayout::NCHW)
```
please combine conditions at 90 and 95

rrawther commented on rocAL/rocAL/source/pipeline/tensor.cpp:
```diff
@@ -0,0 +1,357 @@
+    if(_roi->size() > 0)
+        _roi->clear();
+    _roi->resize(_batch_size);
+    if(_is_image)
```
I don't think, we need to check this here. Please do a memset() to 0 instead which will be much more efficient

rrawther commented on rocAL/rocAL/source/pipeline/tensor.cpp:
```diff
@@ -0,0 +1,357 @@
+    for(unsigned i = 0; i < _num_of_dims; i++)
+        _data_size *= dims.at(i);
+
+    if(_num_of_dims <= 3)
```
I am not sure if this is a correct statement

fiona-gladwin replied:
The image tensors will always be of 4 dimension or greater considering video sequences also(i.e NCHW, NHWC, NFHWC, NFCHW). Hence if the num of dims is less than 3 we are setting is_image flag to false

rrawther commented on rocAL/rocAL/source/pipeline/tensor.cpp:
```diff
@@ -0,0 +1,357 @@
+
+        for (unsigned i = 0; i < info().batch_size(); i++)
+        {
+            if (width[i] > max_width)
```
do std::min() instead and avoid if else. True for line#171

fiona-gladwin replied:
We would check if the given width exceeds the buffer size and raise a warning hence we have kept as if condition.

rrawther commented on rocAL/rocAL/source/pipeline/tensor.cpp:
```diff
@@ -0,0 +1,357 @@
+{
+    if (_vx_handle)
+        return -1;
+
```
there are lots of empty lines in the code. Please remove unnecessary ones following google coding style

rrawther commented on rocAL/rocAL/source/pipeline/tensor.cpp:
```diff
@@ -0,0 +1,357 @@
+
+int rocALTensor::create(vx_context context)
+{
+    if (_vx_handle)
```
add WRN statement that it has been created already

rrawther commented on rocAL/rocAL/source/pipeline/tensor.cpp:
```diff
@@ -0,0 +1,357 @@
+    if(_info._type != rocALTensorInfo::Type::HANDLE)
+        return 0;
+
+#if ENABLE_HIP
```
Is there an equivalent implementation for OpenCL

fiona-gladwin replied:
In AMD OpenVX extensions we do not have the OpenCL support for tensor augmentations, as per our previous discussions  we have only HIP or HOST support hence we have not added the OpenCL copy implementation.

rrawther reviewed (commented):
Please make requested changes and reformat to stick to google coding style

fiona-gladwin committed:
Reformat code and remove blank lines

fiona-gladwin committed:
Reformat code

fiona-gladwin committed:
Change name from rocAL to rocal in tensor files

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into PR_tensor2

rrawther commented on rocAL/rocAL/include/pipeline/tensor.h:
```diff
@@ -0,0 +1,234 @@
+    }
+    void set_max_dims() {
+        if (_layout != RocalTensorlayout::NONE) {
+            _max_dims.resize(2);  // Since 2 values will be stored in the vector
```
max_dims is confusing. Probably better to rename to what it represent  (width, height?)

fiona-gladwin replied:
The max dims will store the width and height in case of image tensors and frames or channels when it comes to Audio tensors hence we have retained the name max_dims

rrawther replied:
In that case it might be better too rename it as _shape. There can be a tensor with dims (256, 3, 4, 4) where max_dims leads to confusing interpretation

fiona-gladwin replied:
@rrawther I have renamed it as max_shape

rrawther commented on rocAL/rocAL/source/pipeline/tensor.cpp:
```diff
@@ -0,0 +1,289 @@
+    _roi->resize(_batch_size);
+    if (_is_image) {
+        for (unsigned i = 0; i < _batch_size; i++) {
+            _roi->at(i).x1 = 0;
```
tensor roi need not be the image dimensions in tensor. It can be set in each node separately based on roi of the input and output

fiona-gladwin replied:
The tensor ROI will be reset each time the dimension of the tensor is changed.
In case of resize/crop the output tensor dimensions will be different from input  tensor dimensions, hence we use the API to reset the ROI of that tensor  wrt the new dimensions.

rrawther reviewed (commented):
@fiona-gladwin I think the use of max_dims() needs to be removed from here. Tensor is generic and in the node it can be interpreted as batch of images.

fiona-gladwin commented:
@rrawther I have addressed the review comments
The max_dims() is used to store the max width/height or max channels and frames. It is used in order to fetch the max values directly instead of having to check the tensor layout each time and fetch the max values from the tensor dimensions.

rrawther reviewed (commented):
Please address the comments when merging new changes

fiona-gladwin committed:
MInor change
